### DISCUSSION TITLE
how to create SQS manually?

### DISCUSSION THREAD

**Discussion by zilahir** [brefphp/bref#1626] [Support]
Hi! 

I am not using `serverless`, but using `CDK`. I would like to create the `SQS` as part of the stack, but i am not sure how to achieve this. 

Checking the `documentation` of [lift](https://github.com/getlift/lift/blob/master/docs/queue.md) it says: 

```
An SQS queue: this is where messages to process should be sent.
A worker Lambda function: this function processes every message sent to the queue.
```

On the [`bref`](https://bref.sh/docs/frameworks/laravel#laravel-queues]) documentation, where the `lift` plugin is used, it passes the handler, as: 

```
constructs:
    jobs:
        type: queue
        worker:
            handler: Bref\LaravelBridge\Queue\QueueHandler
            runtime: php-81
            timeout: 60 # seconds
```

In my `CDK` stack I am trying to replicate this lambda: 

```typescript
const queueLambda = new lambda.Function(
      this,
      `sqs-queue-worker-${env}`,
      {
        functionName: `gitodo-queue-worker-${env}`,
        handler: "", // what should be the handler? in the bref docs its Bref\LaravelBridge\Queue\QueueHandler which i don't think would work here 
        runtime: lambda.Runtime.PROVIDED_AL2,
        layers: [phpRuntimeLayer], // should be the same layer as for my laravel app?
        code: packagePhpCode("../backend"), // should be the same as for my laravel app?
      }
    );
```

It would also help if I would get help on how to create it manually? That way I would be able to reproduce it in `CDK` too.

**Comment by mnapoli** [comment]
Hi! You may want to look inside the Lift codebase, it uses the CDK internally.

**Comment by zilahir** [comment]
hi @mnapoli! 

Yeah, i've checked it. 

However, there's a little issue.

If i do this in my `CDK`:

```typescript
const queueLambda = new lambda.Function(
      this,
      `queue-worker-${env}`,
      {
        functionName: `queue-worker-${env}`,
        handler: "Bref\LaravelBridge\Queue\QueueHandler", // this is from the bref documentation
        runtime: lambda.Runtime.PROVIDED_AL2,
        layers: [phpRuntimeLayer],
        code: packagePhpCode("../backend"),
       ....
      }
    );
```

Getting similar errors to [this](https://github.com/brefphp/bref/issues/1520), meaning the `Handler` is not found. (Of course the bref is present and installed). 

However, if i change the handler to a custom one: 

```typescript
        handler: "sqshandler.php"
```

There's no such issue in the `Lambda`'s log, it executes successfully. 

So the main question at this point is: How to provide the `Bref`'s `QueueHandler` property, as my `Lambda` handler? 

This might the question that easier to answer @mnapoli maybe? :) 

Thank you in advance,

**Reply by mnapoli** [reply]
For `Bref\LaravelBridge\Queue\QueueHandler` to work you need to use the latest versions of Bref and the Laravel bridge, is that the case?

On a separate note, there are CDK constructs for Bref (brefphp/constructs) if you are interested.

**Comment by zilahir** [comment]
Again, thanks for the reply @mnapoli!

Yes, that's the case! 

```
"bref/bref": "^2.1",
"bref/laravel-bridge": "^2.1",
```

are the version I am currently using. 

And yes, I am also using the `brefphp/constructs`, for the api (webapp), and the cli (artisan). If i am not mistaken, what the `brefphp/constructs` does, is adding the related layers to the function. I am doing that manually for this queue lambda for now. 

And the error is: 

```
Fatal error: Illuminate\Container\EntryNotFoundException: Bref\LaravelBridge\Queue\QueueHandler in /var/task/vendor/laravel/framework/src/Illuminate/Container/Container.php:748
```

**Comment by mnapoli** [comment]
> Fatal error: Illuminate\Container\EntryNotFoundException: Bref\LaravelBridge\Queue\QueueHandler

Maybe you disabled automatic registration of service providers from packages? (https://laravel-news.com/package-auto-discovery)

Because it looks like the Bref Laravel bridge isn't registerd? Why would that class be not found 🤔

**Reply by zilahir** [reply]
it was registered, to be honest i am not sure what the solution was, but eventually it started to work. so all good now.

**Reply by Mesuva** [reply]
Sorry for posting on a closed thread, but Interestingly I'm hitting this error myself. 
I'm finding it in my Cloudwatch Logs for the jobs queue.

Full error message is :

{"message":"Bref\\LaravelBridge\\Queue\\QueueHandler","type":"Illuminate\\Container\\EntryNotFoundException","stackTrace":["#0 \/var\/task\/vendor\/bref\/laravel-bridge\/src\/HandlerResolver.php(46): Illuminate\\Container\\Container->get('Bref\\\\LaravelBri...')","#1 \/var\/task\/vendor\/bref\/bref\/src\/FunctionRuntime\/Main.php(26): Bref\\LaravelBridge\\HandlerResolver->get('Bref\\\\LaravelBri...')","#2 \/opt\/bref\/bootstrap.php(17): Bref\\FunctionRuntime\\Main::run()","#3 {main}"]}
 
I've checked that I'm not disabling automatic package registration.
Sending jobs to the queue doesn't seem to be the issue, it's the processing.

I'm using Lift to create the queue, via:

```
constructs:
  jobs:
    type: queue
    worker:
      handler: Bref\LaravelBridge\Queue\QueueHandler
      runtime: php-81
      timeout: 60 # seconds
```
The only unusual thing I'm doing is that the SQS queue is not my default queue. I'm currently using a database based queue, but then specifically sending certain jobs (emails) to the sqs queue.

A little bit stumped on this one!
